platform/features/table/src/controllers/HistoricalTableController.js

```diff
@@ -35,32 +36,79 @@ define(
          * @param telemetryFormatter
          * @constructor
          */
-        function HistoricalTableController($scope, telemetryHandler, telemetryFormatter) {
+        function HistoricalTableController($scope, telemetryHandler, telemetryFormatter, $timeout) {
+            var self = this;
+
+            this.$timeout = $timeout;
+            this.timeoutHandle = undefined;
+            this.batchSize = BATCH_SIZE;
+
+            $scope.$on("$destroy", function () {
+                clearTimeout(self.timeoutHandle);
+            });
+
             TableController.call(this, $scope, telemetryHandler, telemetryFormatter);
         }
 
         HistoricalTableController.prototype = Object.create(TableController.prototype);
 
         /**
-         * Populates historical data on scope when it becomes available from
-         * the telemetry API
-         */
+        * Populates historical data on scope when it becomes available from
+        * the telemetry API
+        */
         HistoricalTableController.prototype.addHistoricalData = function () {
             var rowData = [],
-                self = this;
+                self = this,
+                telemetryObjects = this.handle.getTelemetryObjects();
 
-            this.handle.getTelemetryObjects().forEach(function (telemetryObject) {
-                var series = self.handle.getSeries(telemetryObject) || {},
-                    pointCount = series.getPointCount ? series.getPointCount() : 0,
-                    i = 0;
+            function processTelemetryObject(offset) {
+                var telemetryObject = telemetryObjects[offset],
+                    series = self.handle.getSeries(telemetryObject) || {},
+                    pointCount = series.getPointCount ? series.getPointCount() : 0;
 
-                for (; i < pointCount; i++) {
-                    rowData.push(self.table.getRowValues(telemetryObject,
-                        self.handle.makeDatum(telemetryObject, series, i)));
+                function processBatch(start, end) {
```

Comment:
This is catching my "too many nested calls" alarm.  Also wondering why we use both $timeout and set/clear timeout. One or the other should be good; not both. 

Additionally, there's a lot of "i might get invalid data so i will treat it as valid and continue to process things even when i should not".  I think you can rewrite this as a method on the prototype, something like:

```javascript
HistoricalTableController.prototype.processTelemetry = function (objects, offset, start, rowData) {
    var telemetryObject = objects[offset];
        series,
        end;
        
    if (!telemetryObject) {
        this.stopLoading();
        return;
    }
    
    series = this.handle.getSeries(telemetryObject) 
    
    if (!series) {
        // TODO: No series, what does this mean -- is this a race?
        // Maybe we should get outta here! PANIC AT THE DISCO
        // ...
        return; // phew.  this might not happen due to timeout cancellation.
    }
    
    end = Math.min(series.getPointCount(), end)
    
    for (;start < end; start++) {
        rowData.push(this.table.getRowValues(telemetryObject,
            this.handle.makeDatum(telemetryObject, series, start)));
    }
    if (end >= pointCount) {
        offset += 1;
        end = 0;
    }
    
    this.timeoutHandle = this.$timeout(this.addTelemetry.bind(this, objects, offset, end));
}
```

Then add historical data becomes quite simple-- it kicks off the recursive 
async processing loop (and kills any previous async processing loop).

```javascript
HistoricalTableController.prototype.addHistoricalData = function () {
    if (this.timeoutHandle) {
        this.timeoutHandle.cancel();
    }
    this.timeoutHandle = this.$timeout(this.processTelemetry.bind(
        this,
        this.handle.getTelemetryObjects(),
        0, 
        0
    ));
};
```